cl/lib/es_signal_processor.py

```diff
@@ -141,28 +176,53 @@ def update_es_documents(
     """
     if created:
         return
+
     changed_fields = updated_fields(instance, es_document)
-    if changed_fields:
-        for query, fields_map in mapping_fields.items():
-            main_objects = main_model.objects.filter(**{query: instance})
-            for main_object in main_objects:
-                main_doc = get_or_create_doc(es_document, main_object)
-                if not main_doc:
-                    return
-                fields_to_update = get_fields_to_update(
-                    changed_fields, fields_map
+    if not changed_fields:
+        return
+
+    for query, fields_map in mapping_fields.items():
+        fields_to_update = get_fields_to_update(changed_fields, fields_map)
+
+        if (
+            es_document is PositionDocument
+            and isinstance(instance, Person)
+            and query == "person"
+        ):
+            update_child_documents_by_query.delay(
+                es_document, instance, fields_to_update, fields_map
+            )
+            continue
+
+        if es_document is PositionDocument and isinstance(
+            instance, (ABARating, PoliticalAffiliation, School)
+        ):
+            related_record = Person.objects.filter(**{query: instance})
+            for person in related_record:
+                update_child_documents_by_query.delay(
+                    es_document,
+                    person,
+                    fields_to_update,
+                    fields_map,
                 )
-                if fields_to_update:
-                    update_document_in_es.delay(
+            continue
+
+        main_objects = main_model.objects.filter(**{query: instance})
+        for main_object in main_objects:
+            main_doc = get_or_create_doc(es_document, main_object)
+            if not main_doc:
+                continue
+            if fields_to_update:
+                update_document_in_es.delay(
+                    main_doc,
+                    document_fields_to_update(
                         main_doc,
-                        document_fields_to_update(
-                            main_doc,
-                            main_object,
-                            fields_to_update,
-                            instance,
-                            fields_map,
-                        ),
-                    )
+                        main_object,
+                        fields_to_update,
+                        instance,
+                        fields_map,
+                    ),
+                )
```

Comment:
There's really three different, major code paths here. Should we put comments in front of each, or pull them into their own functions or maybe use a switch? Something to indicate that we're making big decisions before taking actions?

Reply:
Yes, We should. I'll refactor this code.